<?xml version='1.0' encoding='UTF-8'?>
<root>
 <tabbed_widget parent="main_window" name="Main Window">
  <Tab containers="1" tab_name="tab1">
   <Container>
    <DockSplitter count="2" sizes="0.500566;0.499434" orientation="-">
     <DockSplitter count="4" sizes="0.250356;0.249644;0.250356;0.249644" orientation="|">
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="5.269549" bottom="1.039280" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="/joint_states/rev1/position" color="#1f77b4"/>
        <curve name="/joint_reference/data[0]" color="#d62728"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="10.500000" bottom="-10.500000" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="/joint_reference/data[2]" color="#f14cc1"/>
        <curve name="/joint_states/rev1/velocity" color="#9467bd"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="2.493012" bottom="-0.813731" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="/joint_reference/data[1]" color="#ff7f0e"/>
        <curve name="/joint_states/rev_2/position" color="#bcbd22"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="0.830754" bottom="-1.249531" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="/joint_reference/data[3]" color="#d62728"/>
        <curve name="/joint_states/rev_2/velocity" color="#1f77b4"/>
       </plot>
      </DockArea>
     </DockSplitter>
     <DockSplitter count="4" sizes="0.250356;0.249644;0.250356;0.249644" orientation="|">
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="3.263693" bottom="-2.910073" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="position tracking percentage error" color="#1ac938"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="514604.887481" bottom="-128051.564942" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="velocity tracking percentage error" color="#17becf"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="8349.249186" bottom="-40918.196274" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="rev2 position tracking percentage error" color="#1ac938"/>
       </plot>
      </DockArea>
      <DockArea name="...">
       <plot mode="TimeSeries" style="Lines" flip_x="false" flip_y="false">
        <range top="-86.990200" bottom="-129.401817" right="334.300076" left="329.300213"/>
        <limitY/>
        <curve name="Rev 2 velocity tracking percentage error" color="#ff7f0e"/>
       </plot>
      </DockArea>
     </DockSplitter>
    </DockSplitter>
   </Container>
  </Tab>
  <currentTabIndex index="0"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <parameters time_axis="" delimiter="0"/>
  </plugin>
  <plugin ID="DataLoad MCAP"/>
  <plugin ID="DataLoad ROS2 bags">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
   <selected_topics value=""/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS2 Topic Subscriber">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
   <selected_topics value="/joint_states;/joint_reference"/>
  </plugin>
  <plugin ID="UDP Server"/>
  <plugin ID="WebSocket Server"/>
  <plugin ID="ZMQ Subscriber"/>
  <plugin ID="Fast Fourier Transform"/>
  <plugin ID="Quaternion to RPY"/>
  <plugin ID="Reactive Script Editor">
   <library code="--[[ Helper function to create a series from arrays&#xa;&#xa; new_series: a series previously created with ScatterXY.new(name)&#xa; prefix:     prefix of the timeseries, before the index of the array&#xa; suffix_X:   suffix to complete the name of the series containing the X value. If [nil], use the index of the array.&#xa; suffix_Y:   suffix to complete the name of the series containing the Y value&#xa; timestamp:   usually the tracker_time variable&#xa;              &#xa; Example:&#xa; &#xa; Assuming we have multiple series in the form:&#xa; &#xa;   /trajectory/node.{X}/position/x&#xa;   /trajectory/node.{X}/position/y&#xa;   &#xa; where {N} is the index of the array (integer). We can create a reactive series from the array with:&#xa; &#xa;   new_series = ScatterXY.new(&quot;my_trajectory&quot;) &#xa;   CreateSeriesFromArray( new_series, &quot;/trajectory/node&quot;, &quot;position/x&quot;, &quot;position/y&quot;, tracker_time );&#xa;--]]&#xa;&#xa;function CreateSeriesFromArray( new_series, prefix, suffix_X, suffix_Y, timestamp )&#xa;  &#xa;  --- clear previous values&#xa;  new_series:clear()&#xa;  &#xa;  --- Append points to new_series&#xa;  index = 0&#xa;  while(true) do&#xa;&#xa;    x = index;&#xa;    -- if not nil, get the X coordinate from a series&#xa;    if suffix_X ~= nil then &#xa;      series_x = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_X) )&#xa;      if series_x == nil then break end&#xa;      x = series_x:atTime(timestamp)&#x9; &#xa;    end&#xa;    &#xa;    series_y = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_Y) )&#xa;    if series_y == nil then break end &#xa;    y = series_y:atTime(timestamp)&#xa;    &#xa;    new_series:push_back(x,y)&#xa;    index = index+1&#xa;  end&#xa;end&#xa;&#xa;--[[ Similar to the built-in function GetSeriesNames(), but select only the names with a give prefix. --]]&#xa;&#xa;function GetSeriesNamesByPrefix(prefix)&#xa;  -- GetSeriesNames(9 is a built-in function&#xa;  all_names = GetSeriesNames()&#xa;  filtered_names = {}&#xa;  for i, name in ipairs(all_names)  do&#xa;    -- check the prefix&#xa;    if name:find(prefix, 1, #prefix) then&#xa;      table.insert(filtered_names, name);&#xa;    end&#xa;  end&#xa;  return filtered_names&#xa;end&#xa;&#xa;--[[ Modify an existing series, applying offsets to all their X and Y values&#xa;&#xa; series: an existing timeseries, obtained with TimeseriesView.find(name)&#xa; delta_x: offset to apply to each x value&#xa; delta_y: offset to apply to each y value &#xa;  &#xa;--]]&#xa;&#xa;function ApplyOffsetInPlace(series, delta_x, delta_y)&#xa;  -- use C++ indices, not Lua indices&#xa;  for index=0, series:size()-1 do&#xa;    x,y = series:at(index)&#xa;    series:set(index, x + delta_x, y + delta_y)&#xa;  end&#xa;end&#xa;"/>
   <scripts/>
  </plugin>
  <plugin ID="CSV Exporter"/>
  <plugin ID="ROS2 Topic Re-Publisher"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS2 Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations>
  <snippet name="Rev 2 velocity tracking percentage error">
   <global></global>
   <function>return ((value-v1)/v1)*100</function>
   <linked_source>/joint_reference/data[3]</linked_source>
   <additional_sources>
    <v1>/joint_states/rev_2/velocity</v1>
   </additional_sources>
  </snippet>
  <snippet name="rev2 position tracking percentage error">
   <global></global>
   <function>return ((value-v1)/v1)*100</function>
   <linked_source>/joint_states/rev_2/position</linked_source>
   <additional_sources>
    <v1>/joint_reference/data[1]</v1>
   </additional_sources>
  </snippet>
  <snippet name="velocity tracking percentage error">
   <global></global>
   <function>return ((value-v1)/v1)*100</function>
   <linked_source>/joint_states/rev1/velocity</linked_source>
   <additional_sources>
    <v1>/joint_reference/data[2]</v1>
   </additional_sources>
  </snippet>
  <snippet name="position tracking percentage error">
   <global></global>
   <function>return ((value-v1)/v1)*100</function>
   <linked_source>/joint_states/rev1/position</linked_source>
   <additional_sources>
    <v1>/joint_reference/data[0]</v1>
   </additional_sources>
  </snippet>
 </customMathEquations>
 <snippets/>
 <!-- - - - - - - - - - - - - - - -->
</root>

